# Single Cell RNA-Seq Workflow

The scripts in this repository perform the basic steps involved in aligning and analyzing single cell RNA-Seq experiments. For convenience, I have separated them into 2 subdirectories: 1) r_scrna, which performs all steps in a standard single cell analysis workflow from within the R environment in an OS agnostic manner (i.e. should work on most systems) or 2) bash_scrna, which uses bash scripts to do alignments. The R based choice is easier to follow, while the bash scripts are more useful in making custom pipelines on a large compute cluster. Please be aware that the scripts here are NOT intended to be a tutorial, only to provide easy access to common single cell sequencing tasks. When I have time I will put together a comprehensive tutorial.

Single cell RNA-Seq analyses are inherently similar to bulk RNA-Seq analyses. A main difference is each cell in a scRNA-Seq experiment is a unique sequencing library. Also, it should be noted that scRNA-Seq workflows are highly dependent on experimental design and what biological processes are being examined. For example, are you focused of cataloging cell diversity in a heterogeneous population, alternative splicing in cell types or detecting differentially expressed genes? Will you be using plate, microfluidic or droplet based methods? Are you using spike-ins, UMIs or neither? Such differences in experiment design and underlying methodologies often result in slightly different analysis workflows.

The steps in the image below cover a semi-generic single cell RNA-seq workflow. As you can see, there can be many steps involved. I tried to include most common steps, but for clarity the workflow is not comprehensive. As the workflow has numerous steps, I grouped related tasks into related concepts/tasks, which are color coded on the chart below. Relevant scripts and code for each stage is stored in a subdirectory listed to the left of the workflow.

![Alt text](https://github.com/ctrhodes/scRNA-seq/blob/master/scRNA-expanded.png?raw=true)

