# scrna-seq
Single Cell RNA-Seq Workflow

The scripts in this repository perform the basic steps involved in aligning and analyzing single cell RNA-Seq experiemnts. For convenience, I have seperated them into 2 subdirectories: 1) r_scrna, which performs all steps in a standard single cell analysis workflow from within the R environment in an OS agnostic manner (i.e. should work on most systems) or 2) bash_scrna, which uses bash scripts to do aligments. The R based choice is easier to follow, while the bash scripts are more useful in making custom pipelines on a large compute cluster. nPlease be aware that these scripts here are NOT intended to be a tutorial, only to provide easy access to common single cell sequencing tasks. When I have time I will put together a comprehensive tutorial.

Single cell RNA-Seq workflows are inherently similar to bulk RNA-Seq wosrkflows. The main difference is each cell in a scRNA-Seq experiment is a unique sequencing library. Also, it should be noted that scRNA-Seq workflows are highly dependent on experimental design and what biological processes are being examined. For example, are you focused of cataloging cell diversity in a heterogenous population, alternative splicing in cell types or detecting differentially expressed genes? Will you be using plate, microfluidic or droplet based methods? Are you using spike-ins, UMIs or neither? Such differences in experiment design and underlying methodologies often result in slightly different analysis workflows.

The steps in the image below cover a semi-generic single cell RNA-seq workflow. As you can see, there can be many steps involved. I tried to include commonly used steps, but for clarity the workflow is not comprehensive. As the workflow has numerous steps, I grouped related tasks into related stages, which are color coded on the chart below. Relevant scripts and code for each stage is stored in a subdirectory listed to the left of the workflow.

Last, a word about aligners. During next-generation sequencing (NGS) experiments, reads produced by a sequencer are often aligned to a reference sequence, such as a known genome. During RNA-Seq and scRNA-Seq experiments, if the organism from which you obtained RNA is well characterized such as human or mouse, the NGS reads are often aligned to a transcriptome using a variety of alginment algorithms. Historically, aligners such as bowtie/tophat or STAR alignment followed by cufflinks transcript assembly have been used. Currently, "psuedo-aligners" such as kallisto, salmon or sailfish are often used to quantify transcript abundance prior to making inferences at the gene level. I will focus on kallisto, as that is my personal preference. However, kallisto, salmon, sailfish, STAR/cufflinks, and tophat/cufflinks  all perform about the same. and 

and are interchangable in most pipelines. Finally, for many biologists learning about bioinformatics the concepts behind pseudo-aligners seem more difficult to grasp than the process of transcript assembly by "true" aligners such as STAR/cufflinks or tophat/cufflinks. To promote the easiest learning of single cell RNA-Seq workflows I have included scripts for both.

![Alt text](https://github.com/ctrhodes/scRNA-seq/blob/master/scRNA-expanded.png?raw=true)

